File: SPOJ/UFPT2015J.cpp
#include <algorithm>
#include <vector>
#include <string>
#include <iostream>
#include <math.h>
#include <stack>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
using namespace std;

string reverse(string a)
{
	string newString = "";
	for(int i=a.length()-1; i >=0; i--)
	{
		newString += a[i];
	}
	return newString;
}

bool isPalindrome(string input)
{
	bool p;
	string half;
	if (input.length() % 2 != 0)
		{
			half = input.substr(0, input.length()/2 + 1);
		}

		
		else
		{
			half = input.substr(0, input.length()/2);
		}

	string rev = reverse(half);
	if (rev == input.substr(input.length()/2, input.length()))
	{
		return true;
	}
	else
	{
		return false;
	}

}

int main()
{
	string input;
	while (cin >> input)
	{
		if (input == "0") break;
		string half;

		if (isPalindrome(input))
		{
			input = "0";
		}
		else
		{
			bool odd = false;
			//odd, even
			if (input.length() % 2 != 0)
			{
				half = input.substr(0, input.length()/2 + 1);
				odd = true;
			}

			
			else
			{
				half = input.substr(0, input.length()/2);
			}

			//left larger
			string reverseHalf = reverse(half);
			//cout << half << endl;
			long longHalf = atol(reverseHalf.c_str());
			string stringSecondHalf = input.substr(input.length()/2, input.length());
			long secondHalf = atol(stringSecondHalf.c_str());
			if (longHalf > secondHalf)
			{
				long ans = longHalf - secondHalf;
				cout << ans << endl;
			}
			else //not equal needs more work
			{
				half[half.length()-1]++;
				string longHalfA = half;
				string revLongHalfA = reverse(longHalfA);
				string inputTwo;
				if (odd)
				{
					inputTwo = longHalfA + revLongHalfA.substr(1);
				}
				else
				{
					inputTwo = longHalfA + revLongHalfA;
				}
				long longFull = atol(inputTwo.c_str());
				long inputLong = atol(input.c_str());
				cout << longHalfA << " " << revLongHalfA << endl;

			}

			//if not increment
		
		}

		

	}
}

File: SPOJ/UFPT2015J.java
import java.util.*;
import java.io.*;
import java.math.*;
class PALIN 
{
	public static String reverse(String s)
	{
		String newString = "";
		for(int i = s.length()-1; i >=0; i-- )
		{
			newString+=s.substring(i, i+1);
		}
		return newString;
	}

	public static void main(String [] args) throws IOException
	{
		Scanner in = new Scanner(System.in);
		while (1)
		{
			String input = in.next();
			if (input == "0") break;
			Boolean odd = false;
			String half;
			
			if (isPalindrome(intput))//already palin
			{
				input = "0";
			}
			else
			{


			while(true)
			{
				if (input.length() % 2 != 0)//odd
				{
					half = input.substring(0, (input.length()/2)+1);
					odd = true;
				}
				else //even
				{
					half = input.substring(0, (input.length()/2));
				}
				
				String replaceHalf = reverse(half);
				String secondHalf = input.substring(input.length()/2, input.length());
				long replaceHalfInt = Long.parseLong(replaceHalf);
				long secondHalfInt = Long.parseLong(secondHalf);
				if (replaceHalfInt.compareTo(secondHalfInt) == 1)
				{
					input = half.substring(0, (input.length()/2))+ replaceHalf;
					break;
				}
				else
				{
					long halfInt = Long.parseLong(half);
					long newHalfInt = halfInt + 1;
					half = newHalfInt.toString();
					String newInput;
					if (odd)
					{
						newInput = input.substring(input.length()/2 +1, input.length());
					}
					else
					{
						newInput = input.substring(input.length()/2, input.length());
					}
					
					String realNewInput = "";
					for(int j=0; j<newInput.length(); j++)
					{
						realNewInput += "0";
					}
					input = half + realNewInput;
					
					//System.out.println(newHalfInt);
				}


				
			}
		}
			System.out.println(input);

		}
		}

	}
}



